### ISSUE TITLE
Outbox messages are store with defferent "EndpointName" and "ActualEndpointName"

### ISSUE THREAD

**Issue by RomanOlegovich** [BEagle1984/silverback#165]
I expect "ActualEndpointName" equals "EndpointName" in the outbox table, but it's defferent (see screanschort).

Here outbound configuration
```
                        .AddOutbound<NotifyPaymentEvent>(endpoint => endpoint
                            .ProduceTo(nameof(NotifyPaymentEvent))
                            .WithKafkaKey<NotifyPaymentEvent>(envelope => envelope.Message.Id)
                            .Configure(producerConfig)
                            .SerializeAsJsonUsingNewtonsoft()
                            .ProduceToOutbox())
```

This seems to be the cause of the error:
```
fail: Silverback.Messaging.Outbound.TransactionalOutbox.OutboxWorker[1077]
      Failed to produce the message stored in the outbox. | endpointName: StudentUserPhoneDuplicateEvent, messageType: *project-name*.DrivingStartingEvent, *project-name*, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null, messageId: 1e563aad-8b99-4daa-a981-f667d507bb8a, unused1: (null), unused2: (null)
      System.InvalidOperationException: No endpoint with name 'StudentUserPhoneDuplicateEvent' could be found for a message of type '*project-name*.DrivingStartingEvent'.
         at Silverback.Messaging.Outbound.TransactionalOutbox.OutboxWorker.GetTargetEndpoint(Type messageType, String endpointName, IServiceProvider serviceProvider)
         at Silverback.Messaging.Outbound.TransactionalOutbox.OutboxWorker.ProcessMessageAsync(OutboxStoredMessage message, ConcurrentBag`1 failedMessages, IOutboxReader outboxReader, IServiceProvider serviceProvider)
```

![image](https://user-images.githubusercontent.com/46935437/165874553-329bf638-f187-48e4-856d-5eb1aaafbf56.png)

**Comment by BEagle1984** [comment]
It looks like a possible bug, yes.

Can you please provide a full project that reproduce the issue? It's hard to figure it out without seeing the entire configuration.

**Comment by RomanOlegovich** [comment]
I wrote a message but forgot to send it.
To be more clarify. 
StudentPhoneDuplicateEvent and NotifyPaymentEvent are generated in one CronJob, which do:
```
1. Generate list of StudentPhoneDuplicateEvent 
            await SyncStudents() => {
                foreach (var addEvent in addEvents)
                    await publisher.PublishAsync(addEvent);
                outBoxDbContext.SaveChanges();
            }

2.
Generate list of NotifyPaymentEvent 
            await SyncPayments() => {
                foreach (var addEvent in addEvents)
                    await publisher.PublishAsync(addEvent);
                outBoxDbContext.SaveChanges();
            }
```

**Comment by RomanOlegovich** [comment]
I can't provide full project, I'm not owner. 
I can cut out all the pieces of code related to Silverback and paste here :smiley:.

**Comment by davimorao** [comment]
Hi, I encountered the same problem in my tests.
Considering that both commands below are configured to use OutBox, follow the steps to reproduce:

**Code:**
`await _command.ExecuteAsync(new ExecuteSampleCommand
                {
                    Guid = Guid.NewGuid(),
                    TransactionId = Guid.NewGuid(),
                    Sample = new ExecuteSampleCommand.SampleDto
                    {
                        Name = "Test",
                        SampleStatus = ESampleStatus.SampleStatusTwo,
                        InternalId = 1
                    }
                });`
`await kafkaManagementDbContext.SaveChangesAsync();`
                
`await _command.ExecuteAsync(new ExecuteSampleCommand2
                {
                    Guid = Guid.NewGuid(),
                    TransactionId = Guid.NewGuid(),
                    Sample = new ExecuteSampleCommand2.SampleDto
                    {
                        Name = "Test",
                        InternalId = 1
                    }
                });`
`await kafkaManagementDbContext.SaveChangesAsync();`


**The error occurs according to the code above, and will also occur if the commit in the context after sending the first and second message**


**Database**
![image](https://user-images.githubusercontent.com/12645048/216710879-ce275e52-5de2-4216-b984-e1c150dcd269.png)

**Comment by BEagle1984** [comment]
Can you provide a working project that reproduces the issue?
I'd be glad to debug it.

**Comment by davimorao** [comment]
Here's the sample project, let me know if I need more info, thanks.

https://github.com/davimorao/poc-silverback-outbox

**Comment by alefcarlos** [comment]
@BEagle1984 are you good ?

Any news about that issue ? Or any advice how could we workaround that behaviour?

**Comment by BEagle1984** [comment]
I unfortunately didn't have time to debug this issue yet.

An obvious workaround would be to not use the outbox but I guess it's not an option for you.

**Comment by alefcarlos** [comment]
@BEagle1984 , dont you use outbox in production ?

**Comment by BEagle1984** [comment]
Not me personally, as I don't have the need in my current projects. But some colleagues do and I never heard of this issue. Maybe different version, different circumstances, etc.

**Comment by cookie-bytes** [comment]
In OutboxProduceStrategyImplementation I think trying to cache the IProducer is causing the issue.
If I change "ProduceAsync" to the following the endpoint is configured correctly in the outbox table.

` public Task ProduceAsync(IOutboundEnvelope envelope)
            {
                Check.NotNull(envelope, nameof(envelope));

                _logger.LogWrittenToOutbox(envelope);

                if (_producer == null || _producer.Endpoint.Name != envelope.Endpoint.Name)
                {
                    _producer = _outboundQueueBroker.GetProducer(envelope.Endpoint);
                }
                return _producer.ProduceAsync(envelope);
            }`

Does this fix sound reasonable?

Thanks

**Comment by BEagle1984** [comment]
I'm very sorry, I didn't have time yet to look deeply into the issue.

Your explanation makes sense and the fix seems indeed reasonable. Pack it into a PR. 🙂

Thanks a lot! 😎

**Comment by cookie-bytes** [comment]
@BEagle1984 I have created a PR, but the pipeline is failing on a test. I can't get this test to fail locally. Any ideas?

**Comment by BEagle1984** [comment]
It's ok, it's probably flaky on the pipeline. I'm trying to make the tests more reliable but most of the improvements have been implemented for the upcoming big release and not back ported to the current one (in master). I'm checking your pr.

**Comment by BEagle1984** [comment]
I'm still trying to figure out what happened with the test but I uploaded a version (v4.1.1) to nuget.org which includes this PR.
Do you guys have the possibility to test it?

**Comment by cookie-bytes** [comment]
Yes, we will update to the new package tomorrow and start our tests. Thank you

**Comment by davimorao** [comment]
> Here's the sample project, let me know if I need more info, thanks.
> 
> https://github.com/davimorao/poc-silverback-outbox


Hello everyone, I did a test regarding the scenario of this example and messages of different types were stored, sent and consumed correctly.

**Comment by BEagle1984** [comment]
I will release this weekend or at the beginning of next week. Did you try the beta @alefcarlos ?

**Comment by alefcarlos** [comment]
@BEagle1984, @davimorao and I work together rs

We tried the beta and everything worked good!

@cookie-bytes did you try that?

**Comment by cookie-bytes** [comment]
Yes, tested and working as expected. It has not been through performance testing though.

**Comment by BEagle1984** [comment]
Published v4.1.1 including the fix
